src/components/ik-controller.js

```diff
@@ -0,0 +1,234 @@
+      headQuaternion,
+      leftHand,
+      rightHand,
+      rootToChest,
+      iRootToChest
+    } = this;
+
+    // Camera faces the -Z direction. Flip it along the Y axis so that it is +Z.
+    camera.object3D.updateMatrix();
+    cameraForward.multiplyMatrices(camera.object3D.matrix, flipY);
+
+    headTransform.multiplyMatrices(cameraForward, iMiddleEyeToHead);
+    hips.object3D.position.setFromMatrixPosition(headTransform).add(iHipsToHeadVector);
+
+    cameraYRotation.setFromRotationMatrix(cameraForward, "YXZ");
+    cameraYRotation.x = 0;
+    cameraYRotation.z = 0;
+    cameraYQuaternion.setFromEuler(cameraYRotation);
+    Quaternion.slerp(hips.object3D.quaternion, cameraYQuaternion, hipsQuaternion, this.data.rotationSpeed * dt / 1000);
+    hips.object3D.quaternion.copy(hipsQuaternion);
+
+    headQuaternion.setFromRotationMatrix(headTransform).premultiply(hipsQuaternion.inverse());
+
+    head.object3D.quaternion.copy(headQuaternion);
+    hips.object3D.updateMatrix();
+    rootToChest.multiplyMatrices(hips.object3D.matrix, chest.object3D.matrix);
+    iRootToChest.getInverse(rootToChest);
+
+    if (leftController.object3D.visible) {
+      if (!this.leftHandLastVisible) {
+        leftHand.object3D.scale.set(1, 1, 1);
+        this.leftHandLastVisible = true;
+      }
+
+      leftHand.object3D.matrix.multiplyMatrices(iRootToChest, leftController.object3D.matrix);
+
+      const leftHandControls = leftController.components["hand-controls2"];
+
+      if (leftHandControls) {
+        leftHand.object3D.matrix.multiply(leftHandControls.getControllerOffset());
+      }
+
+      leftHand.object3D.matrix.multiply(this.leftHandRotation);
+
+      leftHand.object3D.position.setFromMatrixPosition(leftHand.object3D.matrix);
+      leftHand.object3D.rotation.setFromRotationMatrix(leftHand.object3D.matrix);
+    } else {
+      if (this.leftHandLastVisible) {
+        leftHand.object3D.scale.set(0.0000001, 0.0000001, 0.0000001);
+        this.leftHandLastVisible = false;
+      }
+    }
+
+    if (rightController.object3D.visible) {
+      if (!this.rightHandLastVisible) {
+        rightHand.object3D.scale.set(1, 1, 1);
+        this.rightHandLastVisible = true;
+      }
+      rightHand.object3D.matrix.multiplyMatrices(iRootToChest, rightController.object3D.matrix);
+
+      const rightHandControls = rightController.components["hand-controls2"];
+
+      if (rightHandControls) {
+        rightHand.object3D.matrix.multiply(rightHandControls.getControllerOffset());
+      }
+
+      rightHand.object3D.matrix.multiply(this.rightHandRotation);
+
+      rightHand.object3D.position.setFromMatrixPosition(rightHand.object3D.matrix);
+      rightHand.object3D.rotation.setFromRotationMatrix(rightHand.object3D.matrix);
+    } else {
+      if (this.rightHandLastVisible) {
+        rightHand.object3D.scale.set(0.0000001, 0.0000001, 0.0000001);
+        this.rightHandLastVisible = false;
+      }
+    }
+
+    if (head.object3D.visible) {
```

Comment:
The visible property on the left and right controllers is applied to the left and right hand bones. `ik-root` exists so that we can select the left and right controllers and left and right hands. If we had parent selectors I may be able to use `bone-visibility`. Have any ideas? I do agree that we may not always want to hide your hands on all avatars and probably should be a separate component.